cunumeric/module.py

```diff
@@ -2448,6 +2448,52 @@ def flip(m: ndarray, axis: Optional[NdShapeLike] = None) -> ndarray:
 # Generating index arrays
 
 
+@add_boilerplate("condition", "arr")
+def extract(condition: ndarray, arr: ndarray) -> ndarray:
+    """
+
+    Return the elements of an array that satisfy some condition.
+
+    Parameters
+    ----------
+    condition : array_like
+        An array whose nonzero or True entries indicate the elements
+        of `arr` to extract.
+    arr : array_like
+        Input array of the same size as `condition`.
+
+    Returns
+    -------
+    result : ndarray
+        Rank 1 array of values from arr where `condition` is True.
+
+    See Also
+    --------
+    numpy.extract
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+    """
+
+    if condition.size != arr.size:
+        raise ValueError("arr array and condition array must be of same size")
+
+    if condition.shape != arr.shape:
+        condition_reshape = reshape(condition, arr.shape)
+    else:
+        condition_reshape = condition
+
+    if condition_reshape.dtype == np.dtype(np.bool_):
+        thunk = arr._thunk.get_item(condition_reshape._thunk)
+    else:
+        nonzero_indices = nonzero(condition_reshape)
```

Comment:
I think converting `condition_reshape` to a boolean array and passing it to `get_item` would be more efficient than this code, as `get_item` uses a special compaction task for boolean indices that doesn't materialize coordinates. @ipdemes can you confirm my comment?

Reply:
@magnatelee you are right: replacing `nonzero_indices = nonzero(condition_reshape)` with `bool_condition = condition_reshape.astype(bool)` and then passing `bool condition to arr.get_item will be more efficient (it will skip indirect copy and potentially allocate less memory for temporary data)

Reply:
That sounds like a better approach indeed. I adjusted the code accordingly. Unfortunately there seems to be another issue with `astype(bool)` for complex values that needs to be solved now #548 .

@ipdemes , FYI I tested merging with your branch #504 in favor of my workaround and the problematic testcases are green.

Reply:
@mfoerste4 : do you plan to look at #548 yourself or should it be someone else?

Reply:
@ipdemes @mfoerste4 #549 should fix #548

Reply:
> @mfoerste4 the numpy doc says `np.place` is the exact opposite of `np.extract`, so I guess you could write something similar to this code for `np.extract`. (maybe calling `set_item` instead of `get_item` would suffice.)

Thanks @magnatelee for already fixing #548. I will have a look at np.place in a separate PR